### ISSUE TITLE
Constant cpu usage at 15-20%

### ISSUE THREAD

**Issue by fliegentoeter** [rilian-la-te/vala-panel-appmenu#174]
Hi,

since the last update I have a contant cpu usage of 15-20% from "panel-9-appmenu".

Is there a way to fix this? Has not been this way before.

I am on Manjaro (17.1.1) XFCE (4.12) and are using the xfce-panel with: Window Header - Buttons 0.4.4, Window Header - Title 0.4.4 and the latest vala-panel.

Is anyone else having this problem?

**Comment by rilian-la-te** [comment]
Update to last version and install registrar, please.

**Comment by rilian-la-te** [comment]
And why are you used window header - title?

**Comment by rkunschke** [comment]
Have the problem too now with xfce panel windowck and appmenu

**Comment by rilian-la-te** [comment]
Revision, version?

**Comment by rilian-la-te** [comment]
And please test it heavily. When it happens, what applications give menus when CPU eating starts?

**Comment by rkunschke** [comment]
it eaven heappens in idle with no applications opened.....

**Comment by rkunschke** [comment]
yaourt -Q |grep vala
extra/vala 0.38.5-1
local/vala-panel-appmenu-budgie-git 0.6.1.r63.g919686a-1
local/vala-panel-appmenu-mate-git 0.6.1.r63.g919686a-1
local/vala-panel-appmenu-registrar-git 0.6.1.r63.g919686a-1
local/vala-panel-appmenu-translations-git 0.6.1.r63.g919686a-1
local/vala-panel-appmenu-valapanel-git 0.6.1.r63.g919686a-1
local/vala-panel-appmenu-xfce-git 0.6.1.r63.g919686a-1
local/vala-panel-git 0.3.72.r1.g307d3be-1

**Comment by rilian-la-te** [comment]
Almost latest git, OK. Can you kill an appmenu process in xfce4-panel and run it form scratch. Try to open some applications and see what happens with CPU. It is very strange.

**Comment by rilian-la-te** [comment]
I will look for it in Valgrind this evening along with another Importer issues.

**Comment by fliegentoeter** [comment]
For me this bug went away with the latest updates I installed from AUR.

For the developer: I use window header - title to display the whole window title, since vala does only show the name of the windows and nothing else. Screenshot to show why I use it:


![2018-01-28-192400_1920x1080_scrot](https://user-images.githubusercontent.com/8972985/35485520-3446ed3e-0461-11e8-823d-808d8109fb14.png)

Otherwise I would not have any window titles...

**Comment by rkunschke** [comment]
i tryed with killing and restarting the progress and it changed nothing

**Comment by fliegentoeter** [comment]
Is it possible to show window titles with vala? This would be a nice feature and I could stop using window header title :)

**Comment by rkunschke** [comment]
A bigger menu from application seams to increase the cpu usage

**Comment by rilian-la-te** [comment]
Where is chromium menu? Do you have a Registrar?

**Comment by rilian-la-te** [comment]
Qt or Gtk applications?

**Comment by rilian-la-te** [comment]
Window title - I do not like it, I want application name, not window title)

**Comment by fliegentoeter** [comment]
Vala never worked with chromium or firefox for me.
I am on XFCE (4.12)...

`Window title - I do not like it, I want application name, not window title)`
Ok, but would it be possible to add as an alternative? For people like me this would be awesome!

Here an example with vala working on my machine:
![2018-01-28-193803_1920x1080_scrot](https://user-images.githubusercontent.com/8972985/35485657-1de692c2-0463-11e8-9aa2-b037dfc777c4.png)

**Comment by rkunschke** [comment]
The problem ist in gtk and qt applications

**Comment by rkunschke** [comment]
It some kind of strange but it seams like it about gtk themes.
For instance with materia theme the CPU usage is high.
with numix its nearly on 0 percent

**Comment by rilian-la-te** [comment]
It should work with Chromium, but not with Firefox.
Can you show me a Qt application with menubar?
And do you have vala-panel-appmenu-registrar?

**Comment by rilian-la-te** [comment]
About window title. I should track it, but do not want so (it depends on window manager, so, I do not like it)

**Comment by rilian-la-te** [comment]
I will look for CPU usage cause, and if I find it, I will fix it.

**Comment by fliegentoeter** [comment]
I have installed:
- vala-panel-git
- vala-panel-appmenu-xfce-git

Should I install more?
What does registrar do?

**Comment by rilian-la-te** [comment]
vala-panel-appmenu-registrar-git. If you want appmenu for Chrome or other DBusMenu app, you need Registrar and importer. Importer is inside applet (built as static library), but Registrar from 0.6.0 is standalone and should be installed as separate package.

**Comment by fliegentoeter** [comment]
Is there any up to date instruction on how to get everything working properly?
Haven't found anything about this registrar package so far.

Any need for qt packages to be installed?

**Comment by rilian-la-te** [comment]
No, it is pure glib) It was shown as optional package for your applet)

**Comment by rilian-la-te** [comment]
works for me, so, closing this.

**Comment by stephensrmmartin** [comment]
I'm hitting a worse issue. I'm hitting 72% pretty constantly with appmenu enabled.

![image](https://user-images.githubusercontent.com/2332684/35884412-379a97ce-0b50-11e8-81e7-328c1e241725.png)

See image above.

This is on a fresh reboot; everything freshly installed from git (via the aur package).
```
local/appmenu-gtk-module-git 0.6.1.r125.g332ecf4-1
    Gtk module for exporting menus
local/vala-panel-appmenu-registrar-git 0.6.1.r125.g332ecf4-1
    Gtk module for exporting menus
local/vala-panel-appmenu-translations-git 0.6.1.r125.g332ecf4-1
    Translations for Global Menu
local/vala-panel-appmenu-valapanel-git 0.6.1.r124.g725e54d-1
    AppMenu (Global Menu) plugin for vala-panel
local/vala-panel-appmenu-xfce-git 0.6.1.r125.g332ecf4-1
    AppMenu (Global Menu) plugin for xfce4-panel

```
GTK2 application menus also stuck inside the application [not appearing in global menu] but that's a side issue.

When I kill that process, it restart itself, and instantly continues at ~70% cpu usage.

**Comment by rilian-la-te** [comment]
You do not installed a registrar. So, you need to install it. ```vala-panel-appmenu-registrar-git```

**Comment by stephensrmmartin** [comment]
Yes, it's installed; as I said: local/vala-panel-appmenu-registrar-git 0.6.1.r125.g332ecf4-1

**Comment by rilian-la-te** [comment]
Do it started?

**Comment by rilian-la-te** [comment]
can you start it if not?

**Comment by stephensrmmartin** [comment]
Just by executing /usr/lib/vala-panel/appmenu-registrar ?

If so, I get 
```
(appmenu-registrar:13990): GLib-GIO-CRITICAL **: g_application_release: assertion 'application->priv->use_count > 0' failed
```

According to journalctl, the dbus service started though.

**Comment by stephensrmmartin** [comment]
```
ps -e|grep -i regis
 1636 ?        00:00:00 appmenu-registr
12445 ?        00:00:00 at-spi2-registr
```

**Comment by rilian-la-te** [comment]
it is strange? So, you have CPU usage and gtk2 menus not eaten? Do you have built registrar for gtk2?

**Comment by rilian-la-te** [comment]
and appmenu-gtk-module for gtk2?

**Comment by stephensrmmartin** [comment]
I'm using the AUR PKGBUILD - https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vala-panel-appmenu-xfce-git

I disabled mate, budgie, and vala. Left everything else the same.

**Comment by rilian-la-te** [comment]
Ok, but it is strange then. Do you try with another panel?
Because I need to know, is this XFCE-only or not?

**Comment by stephensrmmartin** [comment]
I'll try. Would the vala panel be a decent test?

**Comment by rilian-la-te** [comment]
Because:
1. I use vala-panel and develop applet primarly for it))
2. It use in-process applets, so, it is easier for debugging AFAIK
3. I want to know if it happens outside of XFCE, and vala-panel has less dependencies in all supported panels)

**Comment by stephensrmmartin** [comment]
From what I can tell, vala-panel is fine. I added a panel, moved it to bottom. Removed all plugins; added the appmenu. CPU usage is normal in general, and for that process.

`/usr/lib/xfce4/panel/wrapper-2.0 /usr/lib/xfce4/panel/plugins/libappmenu-xfce.so 3 12585338 appmenu AppMenu Plugin Display AppMenu (Global Menu)` Uses ~70% of the CPU.

Vala's appmenu hardly uses anything. (< .5%)

**Comment by rilian-la-te** [comment]
It is strange. Do you have another panels (mate and budgie)?

**Comment by stephensrmmartin** [comment]
Ah, wait. I may have figured it out. It seems specific to certain themes.

I like the flatplat family of GTK themes. When those are enabled, the appmenu plugin consumes tons of CPU. When I use simpler themes, the CPU usage problem goes away. I've never had that issue with other plugins. Perhaps you could confirm that problem? I'm uncertain about why this theme would cause the plugin to consume so much CPU.

**Comment by rilian-la-te** [comment]
Can you provide a link to this themes?
I will try it with vala-panel.

And why this themes are so specific? It has a gradients or so?
Do it leaks memory when using this themes?

**Comment by stephensrmmartin** [comment]
It has animations. Runs fine in all other contexts.

Apparently they've updated and changed names to "materia". https://github.com/nana-4/materia-theme

Materia does the same thing as flatplat, with very high CPU usage. I'm uncertain whether it has to do with the theme, per se, or the theming engine (gtk-engine-murrine). 

Edit: Tried another theme that uses the murrine engine, and it was fine. It seems to specifically be something Materia does [again, perhaps animations].

**Comment by rilian-la-te** [comment]
Confirmed, but only for XFCE.

Appmenu have nothing for murrine, it always use GTK3 (it cannot be compiled on GTK2).

Maybe it is GTK2-3 related in XFCE?

Did sntray also consumes a lot of CPU?

**Comment by stephensrmmartin** [comment]
sntray doesn't have the issue from what I can tell. I installed sntray, enabled it in xfce's panel, then ran indicator-kdeconnect. Negligible cpu usage. No difference between themes.

**Comment by stephensrmmartin** [comment]
Hey @rilian-la-te , I think I've found a workaround. If you set the panel to use a solid color instead of the system theme, the CPU usage is back down to normal, regardless of the theme.

That is -> In panel preferences, go to appearance -> style -> Solid color
And voila, it works fine.

**Comment by rilian-la-te** [comment]
Yes, it will work.
